Remove hidden data before validation and make clean data available in DataProcessor

ivarne opened the pull request:
Now all data is cleaned for all (backend) validation if `AppSettings.RemoveHiddenData == true` and `IInstanceDataAccessor` has a new method `GetCleanDataAccessor` so that a DataProcessor (or UserAction) can easily get a copy of the data with hidden data removed for calculations or similar.

One key factor here is that we add a source generator to be able to work with data models more efficiently

During self review I tried to split the PR into smaller commits that stands on its own


* 63018b3c313d749e2b5a55  Move Altinn.App.Analyzer from .Core to .Api
* ecf795db94 Make changes for more efficient removal of hidden data
     * Introduce IFormDataWrapper
     * Deprecate DataModelWrapper
     * New helpers in IInstanceDataAccessor
* 1f015ef1d9fb Add soure generator for IFormDataWrapper implemenations in .Analyzers
    This includes two test projects.
    * SourceGenerator.Tests: Use the generator directly and verify text output
    * SourceGenerator.Integration.Tests: Reference the generator in build process and test the generated code
    
    This has to be two projects, because it is really hard to debug a source generator when it runs as part of the build process.
* 67d2df55c3 Add Clean and Previous InstanceDataAccessors
   This enables DataProcessors to get the clean versions of data models for calculations.
* df07e13cfcb Clean data before validation if AppSettings.RemoveHiddenData = true
* 6589b33747 Minor fixes






## Related Issue(s)
- https://github.com/Altinn/app-lib-dotnet/issues/1157

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

ivarne commented:
/publish

ivarne commented:
/publish

ivarne commented:
/publish

ivarne committed:
Move Altinn.App.Analyzer from .Core to .Api

ivarne committed:
Make changes for more efficient removal of hidden data

 * Introduce IFormDataWrapper
 * Deprecate DataModelWrapper
 * New helpers in IInstanceDataAccessor

ivarne committed:
Add soure generator for IFormDataWrapper implemenations in .Analyzers

This includes two test projects.
* SourceGenerator.Tests: Use the generator directly and verify text output
* SourceGenerator.Integration.Tests: Reference the generator in build process and test the generated code

This has to be two projects, because it is really hard to debug a source generator when it runs as part of the build process.

ivarne committed:
Add Clean and Previous InstanceDataAccessors

This enables DataProcessors to get the clean versions of data models for calculations.

ivarne committed:
Clean data before validation if AppSettings.RemoveHiddenData = true

ivarne committed:
Minor fixes

ivarne committed:
Improve test coverage

ivarne committed:
sonar issues

ivarne commented:
/publish

martinothamar commented on src/Altinn.App.Analyzers/IncrementalGenerator/FormDataWrapperGenerator.cs:
```diff
@@ -0,0 +1,232 @@
+        }
+    }
+
+    private static Result<ModelPathNode> CreateNodeThree(
```
nit: spelling `Three` -> `Tree`

ivarne replied:
😄 

Fixed

martinothamar commented on src/Altinn.App.Analyzers/IncrementalGenerator/FormDataWrapperGenerator.cs:
```diff
@@ -0,0 +1,232 @@
+            {
+                if (dataType.Type != JsonType.Object)
+                {
+                    continue;
```
This branch could emit a diagnostic, all the values in `dataTypes` should be objects

ivarne replied:
Yes, I definitely could add more diagnostics. Not sure I want to do that, because diagnostics from incremental source generators is pretty buggy. I wanted to have a diagnostic for not finding `"ClassRef"` in compilation, and added some extra just because they were easy.

martinothamar replied:
Reading up on the whole "diagnostics from incremental generators" topic, it seems like there is no intention to have a create story there. Maybe we should go in the opposite direction then? Drop all diagnostics here and rather implement analyzers if we feel it is necessary?

martinothamar commented on src/Altinn.App.Analyzers/IncrementalGenerator/FormDataWrapperGenerator.cs:
```diff
@@ -0,0 +1,232 @@
+                var classRef = appLogic.GetProperty("classRef");
+                if (classRef?.Type != JsonType.String)
+                {
+                    continue;
```
If `classRef` is non-null and not a string, this is also invalid configuration

ivarne replied:
I think that `classRef` is required to be "string" whenever `appLogic` is not null/undefined.

The question is how thorough we should be for errors that can get caught with json-schema.

martinothamar commented on src/Altinn.App.Analyzers/IncrementalGenerator/FormDataWrapperGenerator.cs:
```diff
@@ -0,0 +1,232 @@
+        context.RegisterSourceOutput(modelPathNodesProvider, GenerateFromNode);
+    }
+
+    private static ImmutableArray<ModelClassOrDiagnostic> ParseModelClassOrDiagnostic(
```
I think the modelling of the resulting type could be improved here given the additional opportunities for diagnostics. Maybe something like:

```csharp
private sealed record ParseModelClassesResult(
    EquatableArray<(string ClassName, Location? Location)> Models, 
    EquatableArray<EquatableDiagnostic> Diagnostics
);
``` 

Or use a record hierarchy/union pattern if it is beneficial to return a list of results (does `SelectMany` enable concurrency?)

ivarne replied:
I don't see how having an object with two lists provides "additional opportunities for diagnostics" compared to a list of the union of the elements in the two lists.

martinothamar commented on src/Altinn.App.Analyzers/ModelPathNode.cs:
```diff
@@ -0,0 +1,74 @@
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ModelPathNode"/> class.
+    /// </summary>
+    public ModelPathNode(
```
It wasn't obvious to me why `cSharpName` and `jsonName` could be empty in some cases. That is for the root node right? Maybe we could have static named constructors like `CreateRoot` and `CreateProperty` or something. Type hierarchy/union pattern is also an option

ivarne replied:
> That is for the root node right?

Yes, it is only for root.

I think the proper solution would be to create two different types. `Node(List<Property>)` and `Property(cSharpName, jsonName, node)`.

(Yes, I checked and that is what PolyType does have two levels)

Do you think clarity is worth the rewrite?

martinothamar replied:
I think it is worth the rewrite at some point. If we do that now as part of this PR or in the future I'll leave for you to decide  😄

martinothamar commented on src/Altinn.App.Analyzers/SourceTextGenerator/AddIndexToPathGenerator.cs:
```diff
@@ -0,0 +1,156 @@
+        builder.Append(
+            $$"""
+
+                private void AddIndexToPathRecursive_{{node.Name}}(
```
Have you considered using one of those indented writers? A wrapper around `StringBuilder` that has state for current indentation

ivarne replied:
I did consider it, and I'll probably regret my choice if I want to indent code recursively, or easily change between tabs/spaces or `\r\n`/`\n`.

I think the code reads better with the current approach, and appending fewer and appending longer strings should be (very insignificantly) more efficient.

martinothamar replied:
I think it works well for larger blocks where you see the indentation differences in the same string, but when it is broken up due to C# loops etc the readability kind of goes out the window for me

martinothamar commented on src/Altinn.App.Analyzers/SourceTextGenerator/AddIndexToPathGenerator.cs:
```diff
@@ -0,0 +1,156 @@
+
+    private static void GenerateRecursiveMethod(StringBuilder builder, ModelPathNode node, HashSet<string> methods)
+    {
+        if (!methods.Add(node.TypeName))
```
Should we use the `INamedTypeSymbol` for this kind of thing instead?

ivarne replied:
I don't see why that would be better, and it breaks the incremental source generator if I cary a reference to the symbol here.

martinothamar commented on src/Altinn.App.Analyzers/SourceTextGenerator/RemoveGenerator.cs:
```diff
@@ -0,0 +1,170 @@
+                {
+                    CSharpName: "AltinnRowId",
+                    JsonName: "altinnRowId",
+                    TypeName: "global::System.Guid",
```
This feels a little brittle. There is a `IsAltinnRowId` which does something similar

ivarne replied:
Yes, this obviously belongs in a helper in `ModelPathNode`.

Fixed

martinothamar commented on src/Altinn.App.Analyzers/SourceTextGenerator/SourceReaderUtils.cs:
```diff
@@ -0,0 +1,92 @@
+                // Arrays have never been supported in data models.
+                // Just ignore them for now.
+                // return (UnwrapNullable(arrayTypeSymbol.ElementType), arrayTypeSymbol);
+                throw new NotSupportedException(
```
Should this function create diagnostics instead of throwing?

ivarne replied:
Yes, probably, but since 8.0.0 we crash at runtime if anyone has an array in their data model, and nobody complains (probably because they generate models in studio.)

martinothamar commented on src/Altinn.App.Analyzers/SourceTextGenerator/SourceReaderUtils.cs:
```diff
@@ -0,0 +1,92 @@
+                    return (UnwrapNullable(collectionTypeSymbol.TypeArguments[0]), namedTypeSymbol);
+                }
+
+                break;
```
Is it always `ICollection<>`? Feels a little brittle. We could build a set of collection symbols in an earlier stage and do set lookups instead

ivarne replied:
The proper thing to do would probably be to use whatever `System.Text.Json` / PolyType does.

I think we should either try to do the simple thing or rely on a library.

martinothamar commented on src/Altinn.App.Analyzers/SourceTextGenerator/SourceTextGenerator.cs:
```diff
@@ -0,0 +1,132 @@
+                {
+                    return _dataModel as T
+                        ?? throw new InvalidCastException(
+                            $"Attempted to cast data model of type {{rootNode.TypeName}} to {typeof(T).FullName}"
```
nit: I saw from a snapshot that this includes `global::` as well, could probably have another variant that doesn't prefix with that

ivarne replied:
Do you have name suggestion? Currently I have
* Name (without global:: and special caracters)
* TypeName (with global::)

What would the third version of the same be named?

Edit: I chose FullName

martinothamar commented on src/Altinn.App.Analyzers/Utils/EquitableDiagnostic.cs:
```diff
@@ -0,0 +1,87 @@
+        var lineSpan = location.GetLineSpan();
+        var mappedSpan = location.GetMappedLineSpan();
+
+        return Location.Create(lineSpan.Path, location.SourceSpan, lineSpan.Span, mappedSpan.Path, mappedSpan.Span);
```
Interesting, what does trimmed mean here?

ivarne replied:
Without a reference to the symbol/compilation so that the "incremental" part of the generator actually works. It does not matter when I try to reference AdditionalFiles with Location.Create in the first place, but I copied the code from PolyType. I opened a PR to suggest my change, but the whole concept of providing diagnostics from incremental source generators seems broken.  https://github.com/eiriktsarpalis/PolyType/pull/163

ivarne committed:
Fix code review feedback

ivarne committed:
Accept improved error message

Backend currently have 3 failiure conditions where a data model for a reference is not found.
* It does not exist in applicationmetadata
* No dataElement has yet been created
* The dataType has maxCount != 1

Currently the backend message is
"DataType non-existing not found in applicationmetadata.json", which is more specific so I modified the shared tests.

ivarne commented:
/publish

ivarne committed:
Use v4 of json schema for $schema annotations in layouts

ivarne committed:
Cleanup

ivarne committed:
Add more tests for the RequiredLayoutValidator to see how it works with CleanDataAccessor when RemoveHiddenData=true

ivarne committed:
Add ShouldRunAfterRemovingHiddenData to IValidator and proxy into IFormDataValidator along with NoIncrementalValidation

ivarne committed:
Add more telemetry to iinstanceDataAccessor

ivarne committed:
Fix bug in AddIndexToPathGenerator

The generated code could not add indexes to the end

ivarne committed:
Store reference to State in Component context and Obsolete DataModel

ivarne committed:
Check rows for null when evaluating hidden expressions

If a List backing a repeating group contains a null, that is highly likely caused by cleaning
the rows

ivarne commented:
/publish

ivarne committed:
Fix conflicting tests

ivarne committed:
Because of the ["text"] function data accessors needs the translation service

ivarne committed:
Merge artifacts

ivarne commented:
/publish

ivarne commented:
Now that https://github.com/Altinn/app-lib-dotnet/pull/1421 is merged, there is no more need for this
